# LinkedIn Data Science Job Postings
### Hudson Wolfe
### 2024-10-01

## About The Dataset
This data set represents the various facets a large number of Data Science job listings found on LinkedIn, <br> such as the company/organization listing, job location, position and desired skillset.

Dataset courtesy of Asaniczka via kaggle: __https://www.kaggle.com/datasets/asaniczka/linkedin-data-engineer-job-postings/data__

### Dataset Shortcomings
* Job description is in no standardized form and is so varied that analysis would be quite complex. <br>
* Skills are also in no standardized form, had to be parsed from the strings, complicating analysis <br>
* As mentioned, first_seen is all the same<br>

## Exploratory Analysis
The aim of this project is more to explore what insights are to be gleaned from the dataset, rather than answer a particular set of questions.<br>
However, considering the data content several umbrella questions spring to mind immediately:<br>

__What are the characteristics of these jobs?__<br>
__Who's doing the hiring?__<br>
__What skillsets are desired for these positions?__<br>

## Data Preparation

Initialized and verified pandas and dataset, as well as Python's regex library to help with string parsing later:

In [291]:
import pandas as pd
import re
postings = pd.read_csv('postings.csv')
postings.head()

job_title                  company  \
0                               Data Engineer 2             Cook Medical   
1                           Staff Data Engineer  Recruiting from Scratch   
2          Senior Data Engineer, Public Company  Recruiting from Scratch   
3          Senior Data Engineer, Public Company  Recruiting from Scratch   
4  Senior Systems Engineer, Azure Data Platform             Cook Medical   

      job_location                                           job_link  \
0  Bloomington, IN  https://www.linkedin.com/jobs/view/data-engine...   
1  Bloomington, IN  https://www.linkedin.com/jobs/view/staff-data-...   
2  Bloomington, IN  https://www.linkedin.com/jobs/view/senior-data...   
3  Bloomington, IN  https://www.linkedin.com/jobs/view/senior-data...   
4  Bloomington, IN  https://www.linkedin.com/jobs/view/senior-syst...   

   first_seen  search_city search_country   job level job_type  \
0  2023-12-17  Bloomington  United States  Mid senior   Onsite   
1  2023-12-17  Bloomington  United States  Mid senior   Onsite   
2  2023-12-17  Bloomington  United States  Mid senior   Onsite   
3  2023-12-17  Bloomington  United States  Mid senior   Onsite   
4  2023-12-17  Bloomington  United States  Mid senior   Hybrid   

                                         job_summary  \
0  Overview\nThe Data Engineer develops, implemen...   
1  This is for a client of Recruiting from Scratc...   
2  This is for a client of Recruiting from Scratc...   
3  This is for a client of Recruiting from Scratc...   
4  Overview\nWe are seeking a talented Azure Clou...   

                                          job_skills  
0  Azure, SQL, NoSQL, SQL Server, Oracle, MongoDB...  
1  Python, Snowflake, Airflow, Kubernetes, Docker...  
2  Python, SQL, Snowflake, Airflow, Kubernetes, D...  
3  TDD, Automation, Continuous delivery, Data eng...  
4                                                NaN

Dropped columns __"job_link"__ and __"first_seen"__: <br>

__"job_link"__ contained the URL for the listing, irrelavent for this analysis. <br>

__"first_seen"__ contained "The date when the listing was first seen," all of which are listed as 2023-12-17, <br>
presumably the date the data was gathered. 

Renamed and reordered remaining columns for clarity:

In [292]:
# Dropping URL to job posting, 'first_seen'
postings = postings.drop(
    labels = ['job_link', 'first_seen'],
    axis=1
)

#Renaming columns
new_columns = {
    'job_title':'Position', 
    'company':'Company', 
    'job_location':'Location',  
    'search_city': 'Searched City',
    'search_country':'Searched Country',
    'job level':'Level',
    'job_type':'Office Type',
    'job_summary':'Description',
    'job_skills':'Skillset'}
postings = postings.rename(
    mapper=new_columns,
    axis=1
)

#Rearranging column order
postings = postings[['Position','Level','Description','Skillset','Company','Office Type','Location','Searched City','Searched Country']]

postings

Position       Level  \
0                                      Data Engineer 2  Mid senior   
1                                  Staff Data Engineer  Mid senior   
2                 Senior Data Engineer, Public Company  Mid senior   
3                 Senior Data Engineer, Public Company  Mid senior   
4         Senior Systems Engineer, Azure Data Platform  Mid senior   
...                                                ...         ...   
6020                        Senior Oracle Data Analyst  Mid senior   
6021                           Data Governance Analyst  Mid senior   
6022                  Energy Data and Forecast Analyst  Mid senior   
6023                             Junior Data Scientist   Associate   
6024  Lead Data Engineer (Python, SQL, Cloud Engineer)  Mid senior   

                                            Description  \
0     Overview\nThe Data Engineer develops, implemen...   
1     This is for a client of Recruiting from Scratc...   
2     This is for a client of Recruiting from Scratc...   
3     This is for a client of Recruiting from Scratc...   
4     Overview\nWe are seeking a talented Azure Clou...   
...                                                 ...   
6020  We are looking for a highly competent and moti...   
6021  Northampton\nAs a Barclays Data Governance Ana...   
6022  Trilogic Recruitment is pleased to present an ...   
6023  Join the Data Revolution! Become a Junior Data...   
6024  We are looking for a Lead Data Engineer suppor...   

                                               Skillset  \
0     Azure, SQL, NoSQL, SQL Server, Oracle, MongoDB...   
1     Python, Snowflake, Airflow, Kubernetes, Docker...   
2     Python, SQL, Snowflake, Airflow, Kubernetes, D...   
3     TDD, Automation, Continuous delivery, Data eng...   
4                                                   NaN   
...                                                 ...   
6020  Data Analysis, Data Mining, Data Visualization...   
6021  Data Governance, Data Analysis, Data Lineage, ...   
6022  Energy Industry Analysis, Cost Analysis, Reven...   
6023  Python, AI, Machine learning, Scikit, PyTorch,...   
6024                                                NaN   

                      Company Office Type  \
0                Cook Medical      Onsite   
1     Recruiting from Scratch      Onsite   
2     Recruiting from Scratch      Onsite   
3     Recruiting from Scratch      Onsite   
4                Cook Medical      Hybrid   
...                       ...         ...   
6020            Talenterprize      Hybrid   
6021                 Barclays      Hybrid   
6022           Energy Jobline      Hybrid   
6023     Fairmont Recruitment      Onsite   
6024        CenterXchange Inc      Hybrid   

                                    Location  Searched City Searched Country  
0                            Bloomington, IN    Bloomington    United States  
1                            Bloomington, IN    Bloomington    United States  
2                            Bloomington, IN    Bloomington    United States  
3                            Bloomington, IN    Bloomington    United States  
4                            Bloomington, IN    Bloomington    United States  
...                                      ...            ...              ...  
6020       Brackley, England, United Kingdom  Milton Keynes   United Kingdom  
6021    Northampton, England, United Kingdom  Milton Keynes   United Kingdom  
6022  Milton Keynes, England, United Kingdom  Milton Keynes   United Kingdom  
6023         Manchester Area, United Kingdom          Wigan   United Kingdom  
6024     Manchester, England, United Kingdom          Wigan   United Kingdom  

[6025 rows x 9 columns]

To analysize the desired skill metrics accross these listings, it was necessary to gather a comprehensive list of the individual skills:

In [293]:
#Changing null entries to string for parsing
non_null = postings.fillna('None Listed')

#Iterating over each entry as a list, adding unique entries

#Each entry 'skill' in the .tolist() is itself a string of the posting's skillset
all_skills = []
for skills in non_null['Skillset'].tolist():

    #Split the string into a list of skills
    for entry in skills.split(', '):
        #Trimming a few specific nonstandard entry beginnings: "#..." and  "* ...", etc
        entry = re.sub('#', '', entry, count=1)
        entry = re.sub('\* ', '', entry, count=1)

        #Space and comma added to help distinguish entries like: "Data Engineering" and "Data Analysis", which may both flag for just "Data"
        entry = " " + entry + ","
        if entry not in all_skills:
            all_skills.append(entry)

#Confirm
skills_frame = pd.DataFrame({'Skill':all_skills})



The list was then used to populate a skill metrics dataframe:

In [294]:
#Skills given as single string rather than as separate entries, couldn't use .value_counts or similar (Perhaps there's a method I'm not aware of)
#As far as the embedded loop goes, I know there are more efficient ways to do this, just haven't practiced their implementation. Please forgive the loading time

#Could eventually increase speed using "Vectorization"

#Iterates over each skill listing. If the skill is found within, a True is appended to create a boolean mask. 
#Mask is then used to get a dataframe containing only listings desiring the skill. Count and percentage are then derived
skill_search = []
skill_count = []
for skill in all_skills:
    for listing in non_null['Skillset'].tolist():
        listing = " " + listing
        if skill in listing:
            skill_search.append(True)
        else: skill_search.append(False)
    #Count the # of total listings for the skill using a boolean mask
    skill_count.append(postings[skill_search].index.size)
    skill_search = []

In [295]:
#Add in counts and percentages
skills_frame.insert(1, '# Positions Desiring', skill_count)

skill_percent = []
for count in skill_count:
    skill_percent.append(count / 6025 * 100)
skills_frame.insert(2, '% Positions Desiring', skill_percent)

#Sort dataframe
skills_frame = skills_frame.sort_values(by='# Positions Desiring', ascending=False)
skills_frame = skills_frame.reset_index()

#Drop below 75th percentile (Consisting of 1 or 2 occurances each)
above_5 = (skills_frame['% Positions Desiring'] >= 5.0)
skills_frame = skills_frame[above_5]

In [296]:
#Repeating the original search process for just the top skills (5% occurance and above)

#Lists for boolean masks and column entries
skill_search = []
ms_count = []
assoc_count = []

#list containing just the top skills:
top_skills = skills_frame['Skill']

for skill in top_skills:
    for listing in non_null['Skillset'].tolist():
        listing = " " + listing
        if skill in listing:
            skill_search.append(True)
        else: skill_search.append(False)
    #Create dataframe containing listings for the given skill
    searched = postings[skill_search]
    
    #Count # of Mid senior listings for the skill
    ms_skill_listings = searched[searched['Level'] == 'Mid senior']
    ms_count.append(ms_skill_listings.index.size)

    #Count # of Associate listings for the skill
    assoc_skill_listings = searched[searched['Level'] == 'Associate']
    assoc_count.append(assoc_skill_listings.index.size)

    #Reset skill_search
    skill_search = []

In [297]:
#Add count, percentage by position level (5076 total Mid senior (84.35%), 949 total Associate (17.75%))
skills_frame['# Mid Senior'] = ms_count

skills_frame['# Associate'] = assoc_count


## Analysis

### What types of jobs are these?

By nature of the dataset, all of these listings are for positions within or related to the discipline of Data Science. <br>
Specifically, this dataset is comprised of both __Mid Senior__ and __Associate__ level listings from the __US__, __UK__, __Canada__ and __Australia__.

In [298]:
level_by_country = postings.groupby(['Searched Country', 'Level']).agg({'Level':'count'})
level_by_country = level_by_country.rename(
    mapper = {'Level':'# Positions'},
    axis=1
)
level_by_country.reset_index(names=['Country','Level'])


Country       Level  # Positions
0        Australia   Associate           58
1        Australia  Mid senior          230
2           Canada   Associate          233
3           Canada  Mid senior          696
4   United Kingdom   Associate          116
5   United Kingdom  Mid senior          661
6    United States   Associate          542
7    United States  Mid senior         3489

Of the listings themselves, they are primarily for Mid Senior positions:

In [299]:
level_breakdown = postings['Level'].value_counts().rename_axis('Position Level').reset_index(name='Count')
level_breakdown['%'] = postings['Level'].value_counts(normalize=True).reset_index(name='%')['%'] * 100
level_breakdown

Position Level  Count          %
0     Mid senior   5076  84.248963
1      Associate    949  15.751037

The bulk of which are to be found in the US:

In [300]:
#Data Pulled visually from existing dataframe
US_Mid_Senior = pd.DataFrame([['United States', 'Mid Senior', 3489, (3489/6025 * 100), (3849/5076 * 100)]], columns=['Country','Level', '# Positions', '% Total Listings', '% Mid Senior Listings'])
US_Mid_Senior

Country       Level  # Positions  % Total Listings  \
0  United States  Mid Senior         3489         57.908714   

   % Mid Senior Listings  
0              75.827423

Regarding each country, the city which yielded the greatest amount of results are as follows:

In [311]:
#The low number of countries in question allowed visual extraction of data from an existing dataframe
countries = ['United States', 'United Kingdom', 'Canada', 'Australia']
top_cities = ['New York City', 'London', 'Toronto', 'Sydney']
pd.DataFrame({'Country':countries, 'City':top_cities})
#UK - London
#US - NYC
#C - Toronto
#A - Sydney

Country           City
0   United States  New York City
1  United Kingdom         London
2          Canada        Toronto
3       Australia         Sydney

These listings may then be further analyzed according to the office type = __Onsite__, __Hybrid__ and __Remote__:

In [302]:
office_count = postings['Office Type'].value_counts().rename_axis('Office Type').reset_index(name='Count')
office_count['%'] = postings['Office Type'].value_counts(normalize=True).reset_index(name='%')['%'] * 100
office_count

Office Type  Count          %
0      Onsite   2817  46.755187
1      Hybrid   2061  34.207469
2      Remote   1147  19.037344

In [303]:
country_office_types = postings.groupby(['Searched Country','Office Type']).agg({'Office Type':'count'})
country_office_types = country_office_types.rename(
    mapper = {'Office Type':'Count'},
    axis=1
)
country_office_types.reset_index(names=['Country','Office'])

Country  Office  Count
0         Australia  Hybrid    148
1         Australia  Onsite     84
2         Australia  Remote     56
3            Canada  Hybrid    294
4            Canada  Onsite    402
5            Canada  Remote    233
6    United Kingdom  Hybrid    369
7    United Kingdom  Onsite    298
8    United Kingdom  Remote    110
9     United States  Hybrid   1250
10    United States  Onsite   2033
11    United States  Remote    748

As is to be somewhat expected, remote listings remain the smallest portion of the total listings, at a count of __1147__ composing approximately __19%__. <br>
Hybrid listings however are somewhat comparable to those strictly on-site, at a count of __2061 : 2817__ or __34% : 47%__, respectively. <br>

Sorted by level, the relative distribution is maintained:

In [304]:
office_by_level = postings.groupby(['Level', 'Office Type']).agg({'Searched Country':'count'})
office_by_level = office_by_level.reset_index()
office_by_level = office_by_level.rename(
    mapper = {'Searched Country':'Count'},
    axis=1
)
#Small dataset allowed for manual calculation of percentages
office_by_level.insert(3, '% Associate Listings', [34.246575,45.205479,20.547945,0,0,0])
office_by_level.insert(4, '% Mid Senior Listings', [0,0,0,34.200158,47.044917,18.754925])
office_by_level


Level Office Type  Count  % Associate Listings  % Mid Senior Listings
0   Associate      Hybrid    325             34.246575               0.000000
1   Associate      Onsite    429             45.205479               0.000000
2   Associate      Remote    195             20.547945               0.000000
3  Mid senior      Hybrid   1736              0.000000              34.200158
4  Mid senior      Onsite   2388              0.000000              47.044917
5  Mid senior      Remote    952              0.000000              18.754925

### Who's doing the hiring?

The listings within this dataset come from a spread of 2565 unique companies, though it should be noted that the vast majority have only one listing, <br> 
the 75th percentile being a count of just two listings. <br>
Of those that have multiple listings, the breakdown is as follows:


__Top 10 for # listings__

In [305]:
company_by_country = postings.groupby(['Searched Country', 'Company']).agg({'Level':'count'})
company_by_country = company_by_country.rename(
    mapper = {'Level':'Count'},
    axis=1
)
company_by_country = company_by_country.reset_index()
company_by_country = company_by_country.sort_values(by='Count', ascending=False)
company_by_country.head(10)



Searched Country                    Company  Count
2319    United States    Recruiting from Scratch    323
940    United Kingdom  Nigel Frank International     80
2502    United States        Steneral Consulting     61
2313    United States                 Railroad19     58
2104    United States         Motion Recruitment     54
1743    United States                    Gigster     51
1519    United States                  Crossover     49
1301    United States                    BDO USA     49
336            Canada        DeRisk Technologies     42
594            Canada        Steneral Consulting     37

__Top company for listings by country__

In [306]:
#Determined visually from the above table
top_companies = company_by_country.loc[[2319, 940, 336, 160]]
top_companies

Searched Country                    Company  Count
2319    United States    Recruiting from Scratch    323
940    United Kingdom  Nigel Frank International     80
336            Canada        DeRisk Technologies     42
160         Australia        Steneral Consulting     13

In correspondence with it's status as the country with the most listings, the US houses seven of the top ten companies with highest listings. <br>

It's of note that Scratch recruitments are the largest count by a wide margin. <br>
Perhaps it would be useful for potential Data Science applicants to keep an eye on this site, though listings favor those US-based, <br>
as they are split between Onsite and Hybrid offices:

In [307]:
from_scratch = (postings['Company'] == 'Recruiting from Scratch')
from_scratch = postings[from_scratch]
from_scratch = from_scratch['Office Type'].value_counts().to_frame().reset_index()
from_scratch.rename(
    mapper = {'Office Type':'Scratch Positions'},
    axis=1
)

Scratch Positions  count
0            Onsite    189
1            Hybrid    152

### Who are they looking for?

__To reiterate:__ Given the lack of standard formatting within the desired skills dataset multiple entries pertaining to the same skill, <br>
"SQL" and "Experience in SQL" for example, are not both evaluated to the same count of occurances. <br>
This will, of course, skew numerical results, but I don't believe it will obscure such insights as the top skills in demand, etc. <br>

Furthermore, considering the abundance of results with a % occurance of less than 5% (The overwhelming majority of entries, 98.89%), <br>
which includes these non-standardized references to an otherwise highly desired skill, these entries were for the purposes of analysis dropped from the dataframe.<br>
The count and percentage of occurances were calculated with the entirety of entries in consideration, and will therefore remain accurate after row elimination. <br>

__Note:__ The comma in each entry is an artifact of the string parsing process. Removal resulted in a large increase in cell run time, so it was left in for brevity.

In [308]:
skills_frame.head(10)

index               Skill  # Positions Desiring  % Positions Desiring  \
0      1                SQL,                  3057             50.738589   
1      9             Python,                  2905             48.215768   
2     15              Spark,                  1262             20.946058   
3     68                AWS,                  1258             20.879668   
4    479        Engineering,                  1143             18.970954   
5     25                ETL,                  1071             17.775934   
6     50   Data Engineering,                   949             15.751037   
7     10          Snowflake,                   948             15.734440   
8   1419           Analysis,                   920             15.269710   
9      0              Azure,                   887             14.721992   

   # Mid Senior  # Associate  
0          2663          394  
1          2535          370  
2          1173           89  
3          1167           91  
4          1043          100  
5           991           80  
6           880           69  
7           868           80  
8           738          182  
9           821           66

Considering just the top ten for now, SQL and Python stand out as the most desired skills by a fair margin, opening the door to about half of all positions. <br> 

Of these ten, only four are conceptually focused: Analysis, Data Engineering, ETL, and an ambiguous "Engineering". <br>
In the context of Data Science listings, Engineering likely means Data Engineering. In fact, Data Engineering itself is the 7th most desired. <br>
Speculating this to be the case, one could combine listings to shift Data Engineering up to 3rd place with a total of 2,092 listings. This remains speculation, however. <br>

It is of note that several instances of cloud computing are present in the top ten: AWS, Snowflake and Azure

Additionally, SQl and Python remain the highest desired skills within either position level. <br>
Regarding Mid Senior level positions, the highest desired essential mirror the top skills in general:

In [309]:
#Top skills for Mid Senior
top_ms_skills = skills_frame[['Skill','# Mid Senior']].sort_values(by='# Mid Senior', ascending=False)
top_ms_skills = top_ms_skills.iloc[:9]
top_ms_skills

Skill  # Mid Senior
0                SQL,          2663
1             Python,          2535
2              Spark,          1173
3                AWS,          1167
4        Engineering,          1043
5                ETL,           991
6   Data Engineering,           880
7          Snowflake,           868
9              Azure,           821

At the Associate level, an emphasis on concenptual understanding appears, alongside a few more particular languages or frameworks:

In [310]:
#Top skills for Associate
top_assoc_skills = skills_frame[['Skill','# Associate']].sort_values(by='# Associate', ascending=False)
top_assoc_skills = top_assoc_skills.iloc[:10]
top_assoc_skills.reset_index()



index            Skill  # Associate
0      0             SQL,          394
1      1          Python,          370
2      8        Analysis,          182
3     15      Management,          171
4     16   Data Analysis,          151
5     11       Analytics,          149
6     43           Excel,          142
7     13         Tableau,          139
8     18              BI,          121
9     20               R,          120

# Summary

Amongst the postings sampled here, Mid senior level positions compose a majority of approximately 84%, 
most of which are to be found in the US, <br> 
which hosts ~58% of all listings and ~76% of Mid Senior listings. <br>

Only a small portion of listing were explicity remote, only ~19%, the rest being On-Site or partially so as hybrid, 
suggesting that a relocation or commute of some sort would be required. <br>
To best take advantage of, say, the US' job offerings, one would want to relocate around New York City, 
around which are the largest concentration of US jobs. <br>

This is somewhat to be expected from a large metropolitan area, and is characteristic of the other sampled countries, 
each of which had the largest concentration surrounding a major city: London in the Uk, Sydney in Australia and Toronto in Canada. <br>

Of the 2656 companies postings, over 75% have only two listings. Of those which have a larger count, Scratch Recruiting in the US is the standout member at 323 unique listings. <br>
The UK's foremost recruiter is Nigel Frank International, DeRisk Technologies for Canada and Steneral Consulting for Australia. 

The skills these companies are most in demand for are by far SQL and Python proficiencies, being found in approximately 50 and 48 percent of listings, respectively. <br>
Within both Mid Senior and Associate levels, these occupy both 1st and 2nd place. Past these, the Mid senior positions put an emphasis on Cloud Computing expertise through servives such as AWS, Snowflake and Azure alongside some particular Data Science concepts such as ETL. Associate level listings desire additional programming familiarity than just Python and SQL, emphasizing familiarity with frameworks such as Excel and Tableau, as well as the language R. Both Mid Senior and Associate listings share an emphasis on experience in Data Engineering and Analysis. <br>
